trajopt_ifopt/src/discrete_collision_constraint.cpp

```diff
@@ -0,0 +1,264 @@
+ * limitations under the License.
+ */
+#include <trajopt_ifopt/constraints/discrete_collision_constraint.h>
+
+TRAJOPT_IGNORE_WARNINGS_PUSH
+#include <tesseract_kinematics/core/utils.h>
+#include <console_bridge/console.h>
+TRAJOPT_IGNORE_WARNINGS_POP
+
+#include <trajopt_ifopt/constraints/collision_utils.h>
+#include <trajopt_ifopt/utils/numeric_differentiation.h>
+
+namespace trajopt
+{
+DiscreteCollisionConstraintIfopt::DiscreteCollisionConstraintIfopt(DiscreteCollisionEvaluator::Ptr collision_evaluator,
+                                                                   GradientCombineMethod gradient_method,
+                                                                   JointPosition::ConstPtr position_var,
+                                                                   const std::string& name)
+  : ifopt::ConstraintSet(1, name)
+  , position_var_(std::move(position_var))
+  , collision_evaluator_(std::move(collision_evaluator))
+  , gradient_method_(gradient_method)
+{
+  // Set n_dof_ for convenience
+  n_dof_ = position_var_->GetRows();
+  assert(n_dof_ > 0);
+
+  bounds_ = std::vector<ifopt::Bounds>(1, ifopt::BoundSmallerZero);
+}
+
+Eigen::VectorXd DiscreteCollisionConstraintIfopt::GetValues() const
+{
+  // Get current joint values
+  Eigen::VectorXd joint_vals = this->GetVariables()->GetComponent(position_var_->GetName())->GetValues();
+
+  return CalcValues(joint_vals);
+}
+
+// Set the limits on the constraint values
+std::vector<ifopt::Bounds> DiscreteCollisionConstraintIfopt::GetBounds() const { return bounds_; }
+
+void DiscreteCollisionConstraintIfopt::FillJacobianBlock(std::string var_set, Jacobian& jac_block) const
+{
+  // Only modify the jacobian if this constraint uses var_set
+  if (var_set == position_var_->GetName())
+  {
+    // Get current joint values
+    VectorXd joint_vals = this->GetVariables()->GetComponent(position_var_->GetName())->GetValues();
+
+    CalcJacobianBlock(joint_vals, jac_block);
+  }
+}
+
+Eigen::VectorXd DiscreteCollisionConstraintIfopt::CalcValues(const Eigen::Ref<const Eigen::VectorXd>& joint_vals) const
+{
+  Eigen::VectorXd err = Eigen::VectorXd::Zero(1);
+
+  // Check the collisions
+  tesseract_collision::ContactResultVector dist_results;
+  collision_evaluator_->CalcCollisions(joint_vals, dist_results);
+
+  if (dist_results.empty())
+    return err;
+
+  switch (gradient_method_)
+  {
+    case GradientCombineMethod::SUM:
+    {
+      for (tesseract_collision::ContactResult& dist_result : dist_results)
+      {
+        double dist = collision_evaluator_->GetCollisionConfig().collision_margin_data.getPairCollisionMargin(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        double coeff = collision_evaluator_->GetCollisionConfig().collision_coeff_data.getPairCollisionCoeff(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        err[0] += std::max<double>(((dist - dist_result.distance) * coeff), 0.);
```

Comment:
I think the sign needs to be flipped here.

Also, I'm not sure the weighted sum makes sense. Penetration distances greater than the dist setting (currently negative) will still result in an error. I think we want to take the max before the square. Also, I would probably rename it to squared sum or something. 

|Dist | Dist result | error | std::max(error, 0) | std::max(error^2,0) | 
| --------------- | --------------- | --------------- | --------------- | --------------- | 
| 1 | 1.1 | -0.1 | 0 | 0.01|
| 1 | 0.5 |  0.5 | 0.5 | .25 |
| 1 | -0.5 | 1.5 | 1.5 | 2.25 | 
| 0 | 0.5 | -0.5 | 0 | .25 |
| 0 | -0.5 | 0.5 | 0.5 | .25 |